Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs

```diff
@@ -0,0 +1,310 @@
+                userList.Add((GetName(component,nameEntry),nameEntry));
+            }
+
+            userList.Sort(delegate((string,string) a, (string,string) b)
+            {
+                    return String.Compare(a.Item2,b.Item2);
+            });
+
+            state = new MessagesUiState(MessagesUiStateMode.UserList, userList, null);
+        }
+        else
+        {
+            List<MessagesMessageData> messageList = new();
+
+            foreach (var message in component.Messages)
+            {
+                if ((message.SenderId == component.ChatUid && message.ReceiverId == component.UserUid) || (message.ReceiverId == component.ChatUid && message.SenderId == component.UserUid))
+                {
+                    messageList.Add(message);
+                }
+            }
+
+            messageList.Sort(delegate(MessagesMessageData a, MessagesMessageData b)
+            {
+                return TimeSpan.Compare(a.Time,b.Time);
+            });
+
+            List<(string, string)> formattedMessageList = new();
+
+            foreach (var message in messageList)
+            {
+                string name = GetName(component, message.SenderId);
+                var stationTime =message.Time.Subtract(_gameTicker.RoundStartTimeSpan);
+                string content = stationTime.ToString("\\[hh\\:mm\\:ss\\]")+" : "+message.Content;
+                formattedMessageList.Add((name , content));
+            }
+
+            state = new MessagesUiState(MessagesUiStateMode.Chat, formattedMessageList, GetName(component, component.ChatUid));
+        }
+        _cartridgeLoaderSystem?.UpdateCartridgeUiState(loaderUid, state);
+    }
+
+    public void ServerToPdaMessage(EntityUid uid, MessagesCartridgeComponent component, MessagesMessageData message, EntityUid pdaUid, MessagesServerComponent server)
+    {
+        //_sawmill.Debug("Message received - cart");
+        component.Messages.Add(message);
+
+        string name = GetName(component, message.SenderId);
+
+        var wrappedMessage = Loc.GetString("chat-radio-message-wrap",
+            ("color", Color.White),
+            ("fontType", "Default"),
+            ("fontSize", 12),
+            ("verb", "sends"),
+            ("channel", "[PDA]"),
+            ("name", name),
+            ("message", message.Content));
+
+        var chat = new ChatMessage(
+            ChatChannel.Radio,
+            message.Content,
+            wrappedMessage,
+            NetEntity.Invalid,
+            null);
+
+        var chatMsg = new MsgChatMessage { Message = chat };
+
+        if (TryComp(Transform(pdaUid).ParentUid, out ActorComponent? actor))
+            _netMan.ServerSendMessage(chatMsg, actor.PlayerSession.Channel);
+        if (TryComp(pdaUid, out RingerComponent? ringer))
+            _ringer.RingerPlayRingtone(pdaUid, ringer);
+
+
+        if (TryComp(pdaUid, out CartridgeLoaderComponent? loaderUid))
+            UpdateUiState(uid, pdaUid, component);
+    }
+
```

Comment:
You should not be hardcoding in a check for the PDA's parent because this breaks in many situations.